File: python/compress_zip/softmax_cdf.py
"""
Softmax CDF builder for arithmetic coding.

Converts int32 logits to cumulative frequency table for arithmetic coder.
Matches CUDA softmax_cumfreq.cu exactly (softmax_cumfreq_i32_u32_kernel).
"""

import numpy as np
from typing import Tuple

from .primitives import argmax_deterministic
from .lut import get_exp2_lut

# Constants matching CUDA softmax_cumfreq.cu
NUM_STATE_BITS = 32
QUARTER_RANGE = 1 << (NUM_STATE_BITS - 2)  # 1 << 30 = 1073741824

# Exp coefficient matching CUDA (W_CLIP = 0.1, same as lm_head weight clip)
LOG2E = 1.4426950408889634
W_CLIP = 0.1  # Must match lm_head weight clip in model
COEF_D = (W_CLIP * LOG2E) / 64.0
COEF_Q24 = int(COEF_D * (1 << 24) + 0.5)  # ~37817


def compute_target_total(vocab_size: int) -> int:
    """
    Compute target total for frequency table.
    Matches CUDA: target_total = QUARTER_RANGE - vocab_size - 1024
    """
    target = QUARTER_RANGE - vocab_size - 1024
    # Safety check
    if target <= vocab_size:
        target = vocab_size + 1
    return target


def exp_q16_from_diff_acc(diff: int, exp_lut) -> int:
    """
    Compute exp2 for logit difference (int32 GEMM accumulator).
    Matches CUDA exp_q16_from_diff_acc exactly.

    diff = logit - max_logit (<= 0)
    Returns uint16 Q16 representation.
    """
    if diff >= 0:
        return 65535

    neg = -diff

    # t256 = round(neg * COEF_Q24 / 2^24)
    t256 = (neg * COEF_Q24 + (1 << 23)) >> 24

    ip = t256 >> 8  # integer part
    frac = t256 & 255  # fractional part

    if ip >= 31:
        return 1

    base = (1 << 16) >> ip

    # Get fractional multiplier from LUT
    frac_mul = int(exp_lut[frac])

    # Q16 * Q16 -> Q16 (rounded)
    out = (base * frac_mul + 0x8000) >> 16

    if out < 1:
        out = 1
    if out > 65535:
        out = 65535

    return out


def build_cumfreqs(
    logits: np.ndarray,
    vocab_size: int | None = None,
    exp_lut = None,
) -> Tuple[np.ndarray, int]:
    """
    Build cumulative frequency table from int32 logits.
    Matches CUDA softmax_cumfreq_i32_u32_kernel exactly.

    Algorithm:
    1. Find max logit and argmax (smallest index for ties)
    2. Compute exp weights using exp_q16_from_diff_acc
    3. Sum exp weights
    4. Allocate frequencies: freq = 1 + floor(exp * remaining / sum_exp)
    5. Add remainder to argmax
    6. Build inclusive cumsum (no leading 0)

    Args:
        logits: Int32 logits of shape [vocab_size]
        vocab_size: Optional vocab size (defaults to len(logits))
        exp_lut: Optional exp2 LUT (defaults to get_exp2_lut())

    Returns:
        Tuple of (cumfreqs array of shape [vocab_size], target_total)
        cumfreqs is inclusive prefix sum: cumfreqs[-1] = target_total
    """
    if vocab_size is None:
        vocab_size = len(logits)

    assert len(logits) >= vocab_size

    if exp_lut is None:
        exp_lut = get_exp2_lut()
    target_total = compute_target_total(vocab_size)

    # Pass 1: Find max logit
    max_val = int(np.max(logits[:vocab_size]))

    # Find argmax (smallest index for ties) - matches CUDA
    argmax_idx = argmax_deterministic(logits[:vocab_size].tolist())

    # Pass 2: Compute exp weights and sum
    exp_weights = []
    sum_exp = 0

    for i in range(vocab_size):
        diff = int(logits[i]) - max_val  # <= 0
        e = exp_q16_from_diff_acc(diff, exp_lut)
        exp_weights.append(e)
        sum_exp += e

    # Ensure sum_exp is at least 1 to avoid division by zero
    if sum_exp == 0:
        sum_exp = 1

    # Pass 3: Allocate frequencies
    # remaining = target_total - V
    # freq = 1 + floor(exp * remaining / sum_exp)
    remaining = target_total - vocab_size

    freqs = []
    freq_sum = 0

    for i in range(vocab_size):
        e = exp_weights[i]
        # e * remaining needs 64-bit (uint16 * uint32 = up to 2^46)
        add = (e * remaining) // sum_exp
        f = 1 + add
        freqs.append(f)
        freq_sum += f

    # Add remainder to argmax
    rem = target_total - freq_sum
    freqs[argmax_idx] += rem

    # Pass 4: Build inclusive cumsum (no leading 0)
    cumfreqs = np.zeros(vocab_size, dtype=np.uint32)
    running_sum = 0
    for i in range(vocab_size):
        running_sum += freqs[i]
        cumfreqs[i] = running_sum

    # Verify total
    assert cumfreqs[-1] == target_total, f"Total {cumfreqs[-1]} != target {target_total}"

    return cumfreqs, target_total


def build_cumfreqs_simple(
    logits: np.ndarray,
    vocab_size: int | None = None,
) -> Tuple[np.ndarray, int]:
    """
    Simplified CDF builder using floating point softmax then quantizing.

    This is not bit-exact with the integer version but useful for testing.
    """
    if vocab_size is None:
        vocab_size = len(logits)

    target_total = compute_target_total(vocab_size)

    # Floating point softmax
    logits_f = logits[:vocab_size].astype(np.float64)
    logits_f = logits_f - np.max(logits_f)  # Numerical stability
    probs = np.exp(logits_f)
    probs = probs / np.sum(probs)

    # Quantize to frequencies
    freqs = np.round(probs * target_total).astype(np.int32)
    freqs = np.maximum(freqs, 1)  # Minimum freq of 1

    # Adjust to match target_total
    diff = target_total - np.sum(freqs)
    if diff != 0:
        max_idx = np.argmax(freqs)
        freqs[max_idx] += diff

    # Build cumulative frequencies
    cumfreqs = np.zeros(vocab_size + 1, dtype=np.int32)
    cumfreqs[1:] = np.cumsum(freqs)

    return cumfreqs, int(cumfreqs[vocab_size])


# Test cases
if __name__ == "__main__":
    # Test basic CDF building
    logits = np.array([1000, 2000, 3000, 4000], dtype=np.int32)
    cumfreqs, total = build_cumfreqs(logits)

    print(f"Logits: {logits}")
    print(f"Cumfreqs: {cumfreqs}")
    print(f"Total: {total}")

    # Verify properties
    assert cumfreqs[0] == 0
    assert cumfreqs[-1] == total
    assert all(cumfreqs[i] < cumfreqs[i+1] for i in range(len(cumfreqs)-1))

    # Test with uniform logits
    uniform_logits = np.array([1000, 1000, 1000, 1000], dtype=np.int32)
    cumfreqs_uniform, total_uniform = build_cumfreqs(uniform_logits)
    print(f"\nUniform logits: {uniform_logits}")
    print(f"Cumfreqs: {cumfreqs_uniform}")

    # Should be roughly uniform distribution
    freqs = np.diff(cumfreqs_uniform)
    print(f"Frequencies: {freqs}")

    # Test with larger vocab
    large_logits = np.random.randint(0, 10000, 1000, dtype=np.int32)
    cumfreqs_large, total_large = build_cumfreqs(large_logits, 1000)
    print(f"\nLarge vocab (1000): total = {total_large}")
    print(f"Min freq: {np.min(np.diff(cumfreqs_large))}")
    print(f"Max freq: {np.max(np.diff(cumfreqs_large))}")

    print("\nAll softmax CDF tests passed!")

File: rust/src/softmax_cdf.rs
//! Softmax + CDF builder from int32 logits.
//!
//! Converts raw GEMM accumulators (int32) to cumulative frequencies
//! for arithmetic coding.
//! Matches CUDA softmax_cumfreq.cu exactly (softmax_cumfreq_i32_u32_kernel).

use crate::lut::Exp2LutQ16;
use crate::primitives::{argmax_deterministic, round_ties_to_even_host};

/// Arithmetic coder constants
pub const NUM_STATE_BITS: u32 = 32;
pub const STATE_MASK: u64 = (1u64 << NUM_STATE_BITS) - 1;
pub const HALF_RANGE: u64 = 1u64 << (NUM_STATE_BITS - 1);
pub const QUARTER_RANGE: u64 = 1u64 << (NUM_STATE_BITS - 2);

/// Constants matching CUDA softmax_cumfreq.cu
pub const LOG2E: f64 = 1.4426950408889634;
pub const W_CLIP: f64 = 0.1; // Must match lm_head weight clip in model

/// Compute COEF_Q24 matching CUDA exactly.
/// CUDA formula: COEF_D = (W_CLIP * LOG2E) / 64.0
/// where W_CLIP = 0.1 is the lm_head weight clip.
pub fn compute_coef_q24() -> u32 {
    let coef_d = (W_CLIP * LOG2E) / 64.0;
    round_ties_to_even_host(coef_d * (1u64 << 24) as f64).clamp(1, u32::MAX as i64) as u32
}

/// Legacy function for compatibility - use compute_coef_q24() instead.
/// The lm_head_w_clip parameter is ignored; W_CLIP=0.1 is always used.
pub fn compute_coef_q24_for_acc(_lm_head_w_clip: f64) -> u32 {
    compute_coef_q24()
}

/// Compute target_total for frequency allocation.
///
/// Must be < QUARTER_RANGE and leave room for all frequencies being at least 1.
pub fn compute_target_total(vocab_size: usize) -> u32 {
    // target_total = QUARTER_RANGE - vocab_size - 1024
    let target = (QUARTER_RANGE as i64) - (vocab_size as i64) - 1024;
    target.max(vocab_size as i64 + 1) as u32
}

/// Compute exp weight from logit difference using integer LUT.
///
/// Input: diff = logit - max_logit (always <= 0)
/// Output: Q16 approximation of exp(diff * scale)
///
/// This matches the CUDA exp_q16_from_diff_acc function.
#[inline]
pub fn exp_q16_from_diff(diff: i32, exp_lut: &Exp2LutQ16, coef_q24: u32) -> u16 {
    if diff >= 0 {
        return 65535; // exp(0) = 1.0 in Q16
    }

    let neg = (-diff) as u32;

    // t256 = round(neg * coef_q24 / 2^24)
    let t256 = ((neg as u64 * coef_q24 as u64 + (1u64 << 23)) >> 24) as u32;

    let ip = t256 >> 8; // integer part (number of bit shifts)
    let frac = (t256 & 255) as u8; // fractional part (LUT index)

    if ip >= 31 {
        return 1; // Underflow to minimum
    }

    let base = (1u32 << 16) >> ip;
    let frac_mul = exp_lut.lookup(frac) as u32;
    let out = (base * frac_mul + 0x8000) >> 16;

    out.clamp(1, 65535) as u16
}

/// Build cumulative frequencies from int32 logits.
///
/// Input: logits [vocab_size] - raw int32 GEMM accumulators
/// Output: cumfreqs [vocab_size] - cumulative frequencies (inclusive prefix sum)
///
/// The last element of cumfreqs equals target_total.
pub fn build_cumfreqs(
    logits: &[i32],
    exp_lut: &Exp2LutQ16,
    coef_q24: u32,
    target_total: u32,
) -> Vec<u32> {
    let v = logits.len();

    // Step 1: Find max logit and argmax (smallest index for ties)
    let max_val = *logits.iter().max().unwrap_or(&0);
    let argmax_idx = argmax_deterministic(logits);

    // Step 2: Compute exp weights
    let mut exp_weights = Vec::with_capacity(v);
    let mut sum_exp: u64 = 0;

    for &logit in logits {
        let diff = logit - max_val;
        let e = exp_q16_from_diff(diff, exp_lut, coef_q24);
        exp_weights.push(e);
        sum_exp += e as u64;
    }

    // Step 3: Allocate frequencies
    // Each token gets at least 1, plus proportional allocation
    let remaining = target_total as u64 - v as u64;
    let mut freqs = Vec::with_capacity(v);
    let mut freq_sum: u64 = 0;

    for &e in &exp_weights {
        // freq = 1 + floor(e * remaining / sum_exp)
        let extra = (e as u64 * remaining) / sum_exp;
        let freq = 1 + extra as u32;
        freqs.push(freq);
        freq_sum += freq as u64;
    }

    // Step 4: Add remainder to argmax
    let rem = target_total as u64 - freq_sum;
    freqs[argmax_idx] += rem as u32;

    // Step 5: Compute cumulative sum (inclusive prefix sum)
    let mut cumfreqs = Vec::with_capacity(v);
    let mut running_sum: u32 = 0;

    for freq in freqs {
        running_sum += freq;
        cumfreqs.push(running_sum);
    }

    // Verify: last element should equal target_total
    debug_assert_eq!(cumfreqs[v - 1], target_total);

    cumfreqs
}

/// Build cumulative frequencies for a batch of logits.
pub fn build_cumfreqs_batch(
    logits: &[i32], // [B, V] flattened
    vocab_size: usize,
    exp_lut: &Exp2LutQ16,
    coef_q24: u32,
    target_total: u32,
) -> Vec<Vec<u32>> {
    let b = logits.len() / vocab_size;
    let mut result = Vec::with_capacity(b);

    for batch in 0..b {
        let start = batch * vocab_size;
        let row = &logits[start..start + vocab_size];
        result.push(build_cumfreqs(row, exp_lut, coef_q24, target_total));
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_exp_lut() -> Exp2LutQ16 {
        Exp2LutQ16::new()
    }

    #[test]
    fn test_exp_q16_from_diff() {
        let lut = make_exp_lut();
        let coef = compute_coef_q24();

        // diff = 0 should give max (65535)
        assert_eq!(exp_q16_from_diff(0, &lut, coef), 65535);

        // With the small coef (~37819), we need larger diffs to see decay
        let e_1000 = exp_q16_from_diff(-1000, &lut, coef);
        let e_5000 = exp_q16_from_diff(-5000, &lut, coef);
        let e_10000 = exp_q16_from_diff(-10000, &lut, coef);

        // Larger negative diffs should give smaller values
        assert!(e_1000 <= 65535);
        assert!(e_5000 < e_1000, "e_5000={} should be < e_1000={}", e_5000, e_1000);
        assert!(e_10000 < e_5000, "e_10000={} should be < e_5000={}", e_10000, e_5000);
        assert!(e_10000 >= 1, "e_10000={} should be >= 1", e_10000);
    }

    #[test]
    fn test_build_cumfreqs_sum() {
        let lut = make_exp_lut();
        let coef = compute_coef_q24_for_acc(0.1);
        let target_total = compute_target_total(16);

        let logits = vec![100i32, 50, 25, 10, 5, 0, -5, -10, -25, -50, -100, -200, -300, -400, -500, -1000];

        let cumfreqs = build_cumfreqs(&logits, &lut, coef, target_total);

        // Last element should equal target_total
        assert_eq!(cumfreqs[15], target_total);

        // Should be monotonically increasing
        for i in 1..cumfreqs.len() {
            assert!(cumfreqs[i] > cumfreqs[i - 1]);
        }
    }

    #[test]
    fn test_build_cumfreqs_min_freq() {
        let lut = make_exp_lut();
        let coef = compute_coef_q24_for_acc(0.1);
        let target_total = compute_target_total(4);

        // One very high logit, others very low
        let logits = vec![10000i32, -10000, -10000, -10000];

        let cumfreqs = build_cumfreqs(&logits, &lut, coef, target_total);

        // All tokens should have at least freq=1
        assert!(cumfreqs[0] >= 1);
        for i in 1..4 {
            assert!(cumfreqs[i] - cumfreqs[i - 1] >= 1);
        }
    }

    #[test]
    fn test_target_total() {
        let target = compute_target_total(1024);
        assert!(target < QUARTER_RANGE as u32);
        assert!(target > 1024);
    }

    #[test]
    fn test_argmax_gets_remainder() {
        let lut = make_exp_lut();
        let coef = compute_coef_q24_for_acc(0.1);
        let target_total = compute_target_total(4);

        // Identical logits - first one should get remainder (smallest index)
        let logits = vec![100i32, 100, 100, 100];

        let cumfreqs = build_cumfreqs(&logits, &lut, coef, target_total);

        // First token (argmax due to tie-breaking) should have highest freq
        let freq0 = cumfreqs[0];
        let freq1 = cumfreqs[1] - cumfreqs[0];
        assert!(freq0 >= freq1);
    }
}
